### ISSUE TITLE
Discussion on `e` movement and operator-end-inclusiveness

### ISSUE THREAD

**Issue by critiqjo** [martanne/vis#254]
I find Vim's implementation of `e` and `ge` movements idiotic! Consider this (`#` - cursor pos):

```
the quick fox jumped ove# the lazy dog
```

press `dge`

```
the quick fox jump# the lazy dog
```

Contrary to other backward movements, both `r` and `d` were deleted.

`<op>e` is end-inclusive but `<op>w` is end-exclusive. I propose to make `e` and `ge` behave similar to `w` and `b`. That is:

```
the #uick fox jumped over the lazy dog
```

Pressing `e` should result in:

```
the quick#fox jumped over the lazy dog
```

This is also consistent when you have a line cursor -- the cursor will move _after_ the end of the word, and not before that last character of it.

The way it is now, I can't imagine anyone using `ge` along with an operator. And if you are at the `e` of a word (with the current meaning of `e`), you have to do `bce` (or `ciw`) to edit it and not simply `cb`.

I want to know what other users feel about this.

**Comment by critiqjo** [comment]
Speaking of end-exclusivity... To make things consistent, `$` should jump to the trailing `\n` character of the current line and `d$` should be end-exclusive (i.e. exclude the `\n` character).

**Comment by martanne** [comment]
Yes in principle this makes sense. I'm wondering how many people are used stuff like `ea` which would then become `ei`. 

Also I would prefer if we had a proper test suite (see also #166) before fiddling with these core motions. For example the implementation of the word related text objects is currently based on these motions and will likely require some changes.

**Comment by critiqjo** [comment]
This change may cause a few surprises. I wasn't considering symbols in text (because I have a plugin which ignores all the symbols for word movements). Right now, pressing `e` stops at these marked places:

```
myobj.prop=(3); expr;
    ^^   ^ ^^ ^    ^^
```

This proposal suggests to shift all of them right by one character:

```
myobj.prop=(3); expr;
     ^^   ^ ^^ ^    ^^
```

Even though the cursor position changes, the operator behavior with `e` is unchanged by having it be end-exclusive. Nevertheless, the operator behavior with `ge` should be completely changed. First of all, it _should not_ be start-inclusive, and secondly, it _will not_ affect the last character of previous "word".

> I'm wondering how many people are used stuff like `ea` which would then become `ei`.

I'm kinda used to it! :sweat_smile: But I'm angry at `ge`!

> Also I would prefer if we had a proper test suite

I was thinking about separating UI from the engine once and for all. The ability to import the engine as a dynamic library, then run the engine event loop (`vis_run`?) then feeding inputs/events to it and getting updates from it through some channels/callbacks. If this is in place, testing would be a breeze. The more portable (across languages) the library is, the better. Testing is probably done best in a scripting language -- so a Lua support would be great. How much effort do you think this separation would require?

**Comment by critiqjo** [comment]
> How much effort do you think this separation would require?

Well, that question cannot be answered even vaguely without a good idea of what the exposed engine API should be, right? I'll try to open a new issue describing a desirable API, from a mid-high-level view, within a week or two... I'm having a few ideas which seems good to me right now.

**Comment by martanne** [comment]
> I was thinking about separating UI from the engine once and for all.

While this might be desirable for other reasons, it is completely orthogonal to the testing issue.

A working (modulo bugs of course), *simple* testing infrastructure already exists, it just needs somebody to sit down and write *systematic* tests. I would like to stress that coming up with good tests is often non-trivial.

**Comment by critiqjo** [comment]
> A working (modulo bugs of course), simple testing infrastructure already exists

I just tried running it... Except for core tests, it seemed quite slow due to launching a new vis instance for each test case. There should be a faster way to run them. And there is probably an easier way to achieve that without separating the UI; it just suddenly came into my mind...

**Comment by josuah** [comment]
**TL;DR: Just look at the picture and the last line.  :P**

The terminal cursor is a block, and we need to indicate position of
characters with it.

1. `vim` & others embrace the block nature and use it to select a
character.  The position of the cursor indicate the position of a
character (represented as `_`).  The end of the line is never
accessible for that reason: it is not a visible block.

2. `emacs` & others use it to describe a position in between two
characters, like what graphical text editors behavior like acme, sam.
The position of the cursor represents a separation between two
characters (represented below by the `^` that is at the left of the
cursor).  You can go after the last visible character.

3. `vis` uses vim-style where the cursor represent what is
under the block, but it is also possible to access the `\n` character:
it is not a visible block, but still exists, so we can access it like
in emacs and do `x` on it like in vim (or `Ctrl` + `d` in emacs).

```
                        #####             /   _   \
 __       _  |_   .     ##_## __  _   _   |   _)  |   .       _  .,   _   __  .
 |||  \/ (_) |_)  |   . #|_)#|   (_) |_)  |   _)  |   .      (/_ /\  |_) |    .  \n
      /          _|     #|###        |    \       /                  |
^___^___^___^___^___^___#####___^___^___^___^___^___^___^___^___^___^___^___^___^___

                        ^      (1)  we represent the space between the `.` and the `p`
                         ___   (2)  we represent the 'p'
```

when you select from the cursor over the first `p` to cursor over the second `p`:

- `vim`: you selected `prop`, and when you press `e`, the cursor goes to that `p`.
- `emacs`: you selected `pro`, you have to go over the `(` to get `prop` selected.

It seems that you are proposing the second way, which I am fine with.
